libraries/AP_Generator/AP_Generator.h

```diff
@@ -0,0 +1,110 @@
+
+#pragma once
+
+#include <AP_HAL/AP_HAL.h>
+
+#ifndef GENERATOR_ENABLED
+#define GENERATOR_ENABLED !HAL_MINIMIZE_FEATURES
+#endif
+
+#if GENERATOR_ENABLED
+
+#include <AP_Param/AP_Param.h>
+#include <GCS_MAVLink/GCS.h>
+#include <AP_BattMonitor/AP_BattMonitor.h>
+
+class AP_Generator_Backend;
+
+class AP_Generator
+{
+    friend class AP_Generator_Backend;
+    friend class AP_Generator_IE_FuelCell;
+    friend class AP_Generator_RichenPower;
+
+public:
+    // Constructor
+    AP_Generator();
+
+    // Do not allow copies
+    AP_Generator(const AP_Generator &other) = delete;
+    AP_Generator &operator=(const AP_Generator&) = delete;
+
+    static AP_Generator* get_singleton();
+
+    void init(void);
+    void update(void);
+
+    // Return true if AP_Generator is enabled
+    bool enabled() const;
+
+    bool arming_checks(char *failmsg, uint8_t failmsg_len) const;
+
+    AP_BattMonitor::BatteryFailsafe update_failsafes(void);
+
+    // Enum to request reading type
+    enum class ReadingType {
+        voltage,
+        current,
+        consumed_mah,
+        fuel_remain,
+        rpm,
+    };
+
+    // Helpers to retrieve measurements
```

Comment:
The values being returned are in private so that they can't be modified from outside the library.  Is not the correct way of doing this?  What did you have in mind?